Performance: Simpler query for orders months filter dropdown

coreymckrill opened the pull request:
### Changes proposed in this Pull Request:

This introduces an alternative method for generating the list of year-months in the Months filter dropdown. The new method uses a much simpler query that just takes the oldest order and builds the list of options by adding every month from then until the current one. The tradeoff is that if a store has some large gaps of time between some orders, the dropdown may show months where no orders were actually created.

Fixes #54579

### To do

- [x] Maybe update or add unit tests
- [ ] ~Add documentation~

### How to test the changes in this Pull Request:

1. Ensure your test site has HPOS activated, as this change only relates to the list table for HPOS orders.
1. Install [Smooth Generator](https://github.com/woocommerce/wc-smooth-generator/releases/latest) on your test site for easy generation of orders with variable creation dates.
2. Starting with a fresh site with no orders, generate 1 order for the current date: `wp wc generate orders 1`
3. Go to the Orders list table and check the Months filter dropdown. It should just have a single entry for the current month.
4. Now generate an order from 10 years in the past: `wp wc generate orders 1 --date-start=2015-01-01 --date-end=2015-02-01`
7. Check the Months filter dropdown. There should now be an option for each month between 2015 and now.
8. Trash both of the test orders generated so far. Then generate a new one _in the future_ ✨ `wp wc generate orders 1 --date-start=2027-01-01 --date-end=2027-02-01`
9. Check the Months dropdown again. It shouldn't be broken! Instead, it should show every month between now and 2027.

coreymckrill committed:
Perf: Simpler query for orders months filter

This replaces the expensive query that determines every month that an
order was created in with a simple one that just takes the oldest order.
Then we build the list of options by adding every month from then until
the current one. The tradeoff is that if a store has some large gaps of
time between some orders, the dropdown may show months where no orders
were actually created.

Fixes #54579

coreymckrill committed:
Appease the linter

jorgeatorres commented:
> The tradeoff is that if a store has some large gaps of time between some orders, the dropdown may show months where no orders were actually created.

Is this a trade off we're comfortable with? This would deviate from how the same dropdown works in other places in the admin (Posts, Pages, etc.). I understand a store probably has more orders than posts/pages, but still, listing every month since the first order is not too different from a hardcoded "month" selector.

Not saying I'm against, but it's probably worth discussing. Or am I overthinking this?

coreymckrill committed:
Add option to toggle which months method to use

Instead of always using the simplified query for the months filter,
which might trade worse UX for negligible performance gains for stores
with small, sporadic order volume, let's instead keep using the current
query method, and just add a toggle. This way the high volume stores can
enable the simple query as part of their tuning process, and other
stores won't see any change.

This also allows us to make use of the caching from the current method
when using the simplified method.

This also adds a unit test for the simplified method.

coreymckrill committed:
Ensure retrieved option value is boolean

coreymckrill commented:
@jorgeatorres Ok I've refactored this a bit and updated the description and testing instructions. Can you take another look?

joshuatf commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -823,34 +822,116 @@ protected function get_and_maybe_update_months_filter_cache(): array {
 
-		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
-		$order_dates = $wpdb->get_results(
+		$first_year_month = $wpdb->get_row(
```
Nice!  This seems like a reasonable approach.

joshuatf commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -823,34 +822,116 @@ protected function get_and_maybe_update_months_filter_cache(): array {
+					FROM $orders_table t
+					WHERE type = %s
+					AND status != '$trash_status'
+					ORDER BY year ASC, month ASC
```
Just curious if we've tested impact of `ORDER BY` on this query.  I'm not sure that we have alternatives (other than using ID and hoping 🤞  things are chronological), but this clause tends to be one that tends to come up a lot on larger sites.

coreymckrill replied:
Haven't tested it, but I don't think we can rely on the ID being chronological. If a store has ever imported orders from another site or a 3rd party system, they could be from any date but they would still increment the ID column as usual.

joshuatf commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -823,34 +822,116 @@ protected function get_and_maybe_update_months_filter_cache(): array {
+			);
+		}
+
+		update_option( $cache_option_date_name, time(), false );
```
I know this was copied over from the old approach, but curious why this uses option setters over `wp_cache` or even transients with expirations.

joshuatf reviewed (commented):
Sorry for the late review on this one.  I really like where this is headed!  Left a couple comments in the code.

>One way to compromise here, since this change is about optimizing for stores with a large number of orders, might be to add a filter hook here, with the default being the expensive query way, and the filter allowing you to toggle to the more performant way.
>Thoughts, @joshuatf ?

I think we need some way of enabling this by default in larger sites.  My personal opinion is that having more month options to select from is not that much of a nuisance for stores that might not be doing sales every month, but hopefully those cases are in the minority anyway.

If anything, I'd push towards making the simplified the default behavior, but if we feel the need to guard it, we could base the default behavior on store size (number of orders).

>Is this a trade off we're comfortable with? This would deviate from how the same dropdown works in other places in the admin (Posts, Pages, etc.). I understand a store probably has more orders than posts/pages, but still, listing every month since the first order is not too different from a hardcoded "month" selector.

This is a good point, though orders are a very different post type from posts and pages.  They are posts which (ideally) are created in much higher numbers, on a recurring basis, and not by merchants/site-owners.  While I might not appreciate the ability to have to guess months in a filter for posts I've created, it's still valuable information to filter by orders in April of 2024 and find out that no orders exist.

The current dropdown does provide this information as well without the need to navigate to the page, but the question is whether or not this information is beneficial to have on every page load and outweighs performance of loading orders on all other pages/filters.

I would strongly be in favor of performance for most order workflows over recognizing from a dropdown that I did not have orders in a given month from the previous year on every page.

jorgeatorres commented:
> If anything, I'd push towards making the simplified the default behavior, but if we feel the need to guard it, we could base the default behavior on store size (number of orders).

Yes, if we're happy with the less accurate dropdown, I'd prefer we just implement the one approach. Our list table is already overly complex as it is 😬.

> I would strongly be in favor of performance for most order workflows over recognizing from a dropdown that I did not have orders in a given month from the previous year on every page.

Gotcha!

coreymckrill committed:
Revert to original approach

This reverts 606dac73389209aff34435e8937db2c4e5b1fddf and
088ebea2dcbb11f7f36b88a92e6b160f6e825b40 to return to the original
approach of only offering the simplified months filter options and
deprecating the method that runs the expensive query and caches it.

This additionally fixes up the unit tests to work correctly with the
simplified approach.

coreymckrill committed:
Appease the linter

coreymckrill commented:
Ok, based on the conversation above, I've switched back to the original approach of introducing the method with the simplified query and deprecating the old method. @jorgeatorres can you give it another quick look?

coreymckrill committed:
Merge branch 'trunk' into fix/54579/order-month-query

jorgeatorres commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -802,61 +801,93 @@ private function months_filter() {
+					       MONTH( t.date_created_gmt ) AS month
+					FROM $orders_table t
+					WHERE type = %s
+					AND status != '$trash_status'
```
Maybe we can use `%s` here and pass `OrderStatus::TRASH` as an arg instead of including it in the query itself?

coreymckrill replied:
Updated in 6e66ebae0d

jorgeatorres commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -802,61 +801,93 @@ private function months_filter() {
+			$start = new \DateTime(
+				sprintf(
+					'%s-%s-01',
+					$first_year_month->year,
+					$first_year_month->month
```
Note that `$first_year_month` is in UTC (because it comes from `date_created_gmt`) so we can't assume it's in the WP timezone. For example, if your WP timezone is UTC-5 and you have an order with `date_created_gmt = 2025-01-01 01:00:00`, the dropdown will list January 2025, but should in fact list December 2025 (which is the order date in WP's timezone).

coreymckrill replied:
Good catch! I suspect the old approach has this same defect. I updated this in 6e66ebae0d and added a unit test for it.

jorgeatorres commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -802,61 +801,93 @@ private function months_filter() {
+				$start = new \DateTime( 'now', wp_timezone() );
+			}
+
+			while (
```
I might be missing something, but couldn't we use `DatePeriod` instead of all this "manual" adding of intervals?

```php
foreach ( new \DatePeriod( $start, new \DateInterval( 'P1M' ), $end ) as $d ) {
	// ...
}
```

coreymckrill replied:
I actually didn't know about `DatePeriod` 😳 I tried your suggestion, but it doesn't always work quite right because it excludes the end date from the list of recurrences. There's an optional flag `DatePeriod::INCLUDE_END_DATE` to adjust this, but it's not available until PHP 8.2.

jorgeatorres replied:
Yeah, we would have to include the end date manually, but maybe that's still an improvement / or at least more clear? This is not a blocker, of course.

coreymckrill replied:
Just my personal preference, I guess, but I think it's actually clearer/less complex if we don't have to add extra logic in order to get the correct last item. I think I'll leave it for now.

jorgeatorres reviewed (changes requested):
Hey @coreymckrill! Overall this looks good to me and tests well, except for a corner case which I've documented in the feedback. Other than that, I just left a few tiny suggestions. Let me know what you think!

coreymckrill committed:
Account for timezone offset from GMT

Ensure that the month options reflect the months in the site's timezone.
The order date is stored in GMT, so it could end up being a different
month and/or year than the local order date.

Also updates the months query so that the trash status is included in
the SQL statement via the prepare method, rather than interpolation.

coreymckrill committed:
Swap expected and actual on several tests

coreymckrill committed:
Merge branch 'trunk' into fix/54579/order-month-query

jorgeatorres reviewed (approved):
LGTM. Tests well (tried a few corner cases by changing the timezone on my test site as well as the order's creation date).
Thank you @coreymckrill!
